File: app/src/main/java/doubleslash05/mini/team11/ui/sj/RecipeActivity.java
package doubleslash05.mini.team11.ui.sj;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

import com.google.android.material.tabs.TabLayout;

import doubleslash05.mini.team11.R;
import doubleslash05.mini.team11.model.data.Button;
import doubleslash05.mini.team11.model.data.Recipe;
import doubleslash05.mini.team11.model.data.Recipe_Base;
import doubleslash05.mini.team11.model.data.Step;
import doubleslash05.mini.team11.model.data.Text3;

public class RecipeActivity extends AppCompatActivity {

    RecyclerView recyclerView;
    RecipeAdapter recipeAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_recipe);

        setTitle("팬케이크");


        recyclerView = findViewById(R.id.recyclerRecipe);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        recipeAdapter = new RecipeAdapter(this);
        recyclerView.setAdapter(recipeAdapter);

        TabLayout tabs=(TabLayout) findViewById(R.id.tabs);
        /*tabs.addTab(tabs.newTab().setText("재료 & 정보"));
        tabs.addTab(tabs.newTab().setText("레시피"));
        tabs.setTabGravity(tabs.GRAVITY_CENTER);


        final ViewPager viewPager = (ViewPager) findViewById(R.id.viewPager);
        final PageAdapter myPageAdapter = new PageAdapter(getSupportFragmentManager(),2);
        viewPager.setAdapter(myPageAdapter);

        tabs.addOnTabSelectedListener(new TabLayout.ViewPagerOnTabSelectedListener(viewPager));
        viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabs));*/

        Recipe_Base[] list = new Recipe_Base[]{
                new Recipe("주말에는 여유롭게 브런치", "팬케이크", "15분", "초급"),



                new Step("STEP 1", R.drawable.step_empty),
                new Text3("1. 믹스와 물을 잘 섞어줍니다."),
                new Text3("2. 반죽에 계란을 넣고 섞습니다."),
                new Text3("3. 멍울이 지지 않게 풀어줍니다."),

                new Step("STEP 2", R.drawable.step_empty),
                new Text3("1. 달궈진 팬에 식용유를 두르고 약불로."),
                new Text3("2. 반죽을 국자의 반정도 퍼서 동그랗게"),
                new Text3("3. 반죽 표면에 기포가 보글보글 올라오면 뒤집개를 이용해 뒤집어줍니다"),

                new Step("STEP 3", R.drawable.step_empty),
                new Text3("1. 뒤집은 상태에서 약 1분간 기다린후 불을 꺼줍니다."),
                new Text3("2. 시럽을 붓고 버터를 올립니다."),
                new Text3("3. 기호에 따라 과일 등 다양한 토핑을 곁들입니다."),

                new Button(),
        };
        recipeAdapter.setData(list);
    }
}

File: app/src/main/java/doubleslash05/mini/team11/ui/recipe/RecipeActivity.kt
package doubleslash05.mini.team11.ui.recipe

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.MenuItem
import android.widget.CheckBox
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import com.google.android.material.tabs.TabLayout
import com.icaksama.rapidsphinx.RapidPreparationListener
import com.icaksama.rapidsphinx.RapidSphinx
import com.icaksama.rapidsphinx.RapidSphinxListener
import doubleslash05.mini.team11.App
import doubleslash05.mini.team11.R
import doubleslash05.mini.team11.model.RecipeModel
import doubleslash05.mini.team11.model.data.RecipeData
import doubleslash05.mini.team11.model.data.RecipeVideoData
import doubleslash05.mini.team11.model.network.base.ApiStatus
import doubleslash05.mini.team11.ui.base.BaseActivity
import doubleslash05.mini.team11.ui.common.widget.recipevideo.RecipeVideoView
import doubleslash05.mini.team11.util.Log
import edu.cmu.pocketsphinx.Config
import kotlinx.android.synthetic.main.activity_recipe.*
import kotlinx.android.synthetic.main.view_recipe_video.*
import kotlinx.android.synthetic.main.view_recipe_video.view.*

class RecipeActivity : BaseActivity(), RapidSphinxListener, TabLayout.OnTabSelectedListener {
    private val infoFragment = RecipeInfoFragment()
    private val stepFragment = RecipeStepFragment()
    private val menuId by lazy { intent.getIntExtra(EXTRA_MENU_ID, -1) }
    private var data: RecipeData? = null
    private val rapidSphinx: RapidSphinx by lazy { RapidSphinx(this) }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_recipe)

        setSupportActionBar(toolbar_recipe)
        supportActionBar?.setDisplayShowTitleEnabled(false)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        checkbox_recipe_favorite.setOnClickListener {
            val data = data ?: return@setOnClickListener
            it as CheckBox
            RecipeModel.setFavorite(data.id, it.isChecked)
        }

        rapidSphinx.addListener(this)

        button_recipevideo_start.setOnClickListener {
            player_recipevideo.start()
            rapidSphinx.startRapidSphinx(5000)
            rapidSphinx.setSilentToDetect(1000)
        }


        rapidSphinx.prepareRapidSphinx(object : RapidPreparationListener {
            override fun rapidPreExecute(config: Config) {
                rapidSphinx.isRawLogAvailable = true
                config.setString("-logfn", "/dev/null")
                config.setBoolean("-verbose", true)
            }

            override fun rapidPostExecute(isSuccess: Boolean) {
                Log.d("rapid2", "Executed!")
            }
        })

        videoview_recipe.setOnChangeSectionListener(object :
            RecipeVideoView.OnChangeSectionListener {

            override fun onChangeSection(index: Int) {
                stepFragment.setStep(index)
            }
        })

        tablayout_recipe.addOnTabSelectedListener(this)

        if (App.prefs.isShowVoiceGuide)
            GuideVoiceDialog().show(supportFragmentManager, null)

        supportFragmentManager.beginTransaction()
            .add(R.id.fragment_recipe, stepFragment)
            .add(R.id.fragment_recipe, infoFragment)
            .commit()

        RecipeModel.getRecipeDetail(menuId).observe(this, Observer {
            when (it) {
                is ApiStatus.Success -> {
                    data = it.data
                    refresh()
                }
            }
        })


    }

    override fun rapidSphinxDidStop(reason: String?, code: Int) {
        rapidSphinx.stop()
        rapidSphinx.startRapidSphinx(10000)
    }

    override fun rapidSphinxFinalResult(
        result: String?,
        hypArr: MutableList<String>?,
        scores: MutableList<Double>?
    ) {
    }

    override fun rapidSphinxPartialResult(partialResult: String?) {
        when (partialResult) {
            stopKeyword2 -> videoview_recipe.prevSection()
            pauseKeyword -> videoview_recipe.replySction()
            nextKeyword -> videoview_recipe.nextSection()
        }
    }

    override fun rapidSphinxUnsupportedWords(words: MutableList<String>?) {
    }

    override fun rapidSphinxDidSpeechDetected() {
        Log.d("rapid", "detected")
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }

        return super.onOptionsItemSelected(item)
    }

    @Override
    override fun onResume() {
        val oovwords = arrayOf("darcy", "daum", "region")
        super.onResume()
        rapidSphinx.updateVocabulary(stopKeyword, oovwords) {
            Log.d("rapid", "updated!")

        }
    }

    override fun onTabSelected(tab: TabLayout.Tab) {
        val transaction = supportFragmentManager.beginTransaction()
        when (tab.position) {
            0 -> transaction.show(infoFragment).hide(stepFragment)
            1 -> transaction.hide(infoFragment).show(stepFragment)
        }
        transaction.commit()
    }

    override fun onTabReselected(tab: TabLayout.Tab) {

    }

    override fun onTabUnselected(tab: TabLayout.Tab) {

    }

    @SuppressLint("SetTextI18n")
    fun refresh() {
        val data = data ?: return
        textview_recipe_title.text = data.name
        val recipeVideoData = RecipeVideoData(data.videoUrl, data.ms)
        videoview_recipe.setData(recipeVideoData)

        textview_recipe_main.text = data.name
        textview_recipe_sub.text = data.descriptionShort
        textview_recipe_time.text = "${data.cookingTime}분"
        textview_recipe_level.text = data.level

        checkbox_recipe_favorite.isChecked = data.favorite

        infoFragment.setData(data)
        stepFragment.setData(data)
    }

    companion object {
        const val EXTRA_MENU_ID = "EXTRA_MENU_ID"
        const val EXTRA_MENU_DATA = "EXTRA_MENU_DATA"
        const val pauseKeyword = "darcy"
        const val stopKeyword = "stop"
        const val stopKeyword2 = "region"
        const val nextKeyword = "daum"
    }

    private val isPermissionsGranted: Boolean
        private get() = if (ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.RECORD_AUDIO
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this, arrayOf(Manifest.permission.RECORD_AUDIO),
                1
            )
            false
        } else {
            true
        }
}